### DISCUSSION TITLE
Error during morphometrics extraction

### DISCUSSION THREAD

**Discussion by Paasu** [neuropoly/axondeepseg#551] [Help needed] [answered]
When I try to get the morphometrics from my images, I get the following error.
[error.txt](https://github.com/neuropoly/axondeepseg/files/6923151/error.txt)
This occurs when I try to do ellipse or circle, folder or single file.
This was done in Windows 10

**Comment by mathieuboudreau** [comment]
Hi @Paasu,

Could you maybe provide the image or image folder that you are trying to run the morphometrics on? It looks like it may be trying to segment an image that's not an ADS segmentation.

**Reply by Paasu** [reply]
There are about 80-90 images I'm trying to segment so I can't provide the whole folder over this thread but I can give some example ones?
![ID22](https://user-images.githubusercontent.com/50454820/128020839-978f4809-d802-46b0-99e6-9ec0a8b9ec43.png)
![ID24](https://user-images.githubusercontent.com/50454820/128020843-5a005af0-b93c-4a88-9ae8-f3a4e4f5e059.png)
![ID76](https://user-images.githubusercontent.com/50454820/128020863-035c0a2d-6b54-423a-9098-f84efdff367d.png)

![ID22_seg-axonmyelin](https://user-images.githubusercontent.com/50454820/128020762-e017464a-a09c-4fda-812a-488db26de9a4.png)
![ID24_seg-axonmyelin](https://user-images.githubusercontent.com/50454820/128020771-77783350-59f9-4526-9465-bcf01b13b19f.png)
![ID76_seg-axonmyelin](https://user-images.githubusercontent.com/50454820/128020890-f0defae5-f380-4235-adb1-9f9e0f3f9863.png)

I made corrections to the images using Photoshop in areas where segmentation didn't fully pickup

**Reply by mathieuboudreau** [reply]
Perfect, thanks! I'm suspecting your corrections in Photoshop might have caused an issue. Would you mind trying your morphometrics command on a single image that wasn't manually corrected using photoshop (ie. direct output from AxonDeepSeg)?

**Reply by Paasu** [reply]
It appears to output a morphometrics file from the looks of it. The corrections are quite important however, is there anyway I can use those images?

**Reply by hermancollin** [reply]
@Paasu Are the corrected images grayscale PNGs?

**Reply by Paasu** [reply]
I believe so. The myelin is 148 RGB and the rest are white and black.

**Reply by hermancollin** [reply]
Are the segmentations you provided above the manual corrections you mentioned?

**Reply by Paasu** [reply]
Yeah a select sample from a larger dataset with brief manual corrections done to all of them after segmentation by ADS.

**Comment by hermancollin** [comment]
Looking at one of the provided images, I can see some problems that might be causing the issue:
1. The color profile is RGB. I think it should be grayscale. This can be modified in Photoshop before saving.
2. If I'm not mistaken, ADS expects 3 values: black (0) for background, gray (~127) for myelin and white (255) for axon. (Not sure if the 0-127-255 values are actually a requirement. I think ADS expects only 3 values tho) It is possible that your manual corrections introduced some undesired artifacts in the masks. As you can see below, zooming in on one of the masks, we can see that there are more than 3 grayscale values. This can be caused either by too much compression (which introduces some intermediate values in the color profile to reduce the file size) or, most likely, you used a soft brush for the manual correction. These brushes produce "soft edges" with intermediate grayscale values.


![Screenshot_20210803_115922](https://user-images.githubusercontent.com/83031821/128047874-b247bfdc-d050-45e2-8ecb-4bab9a1a40ac.png)

**Reply by hermancollin** [reply]
To fix the "too many grayscale values" problem, I suggest you follow the guidelines in the [morphometrics extraction notebook](https://github.com/neuropoly/axondeepseg/blob/master/notebooks/04-morphometrics_extraction.ipynb).

More specifically, the following code (or something along those lines) could help you threshold your corrections:
```python

# Read paths and compute axon/myelin masks
img = ads.imread(path_img)
pred = ads.imread(path_pred)
pred_axon = pred > 200
pred_myelin = np.logical_and(pred >= 50, pred <= 200)
path_folder, file_name = os.path.split(path_img)
```

**Reply by Paasu** [reply]
I did use a brush for the manual corrections which is probably the cause of the other grayscale values. Switching the image to grayscale doesn't help. I think the other grayscale values are the cause of the problem. I tried using the morphometrics extraction but it doesn't output the excel file but an aggregate of the values instead. I think I might do the corrections again using the pencil tool so it avoids adding other grayscale values.

Thank you very much for your help!

**Reply by mathieuboudreau** [reply]
Hi @Paasu,

Our recommended approach to correct images is using our AxonDeepSeg FSLeyes plugin.

Best,

Mathieu

**Reply by Paasu** [reply]
Unfortunately FSLeyes doesn't seem to work on my Windows or Linux system.
![ID1_seg-myelin](https://user-images.githubusercontent.com/50454820/128307539-f7cebda7-234b-4a40-9fe0-4cd9900b96b5.png)
![ID1_seg-axon](https://user-images.githubusercontent.com/50454820/128307548-c7d07ab0-586c-41a7-9c2b-fc63ffcd88a2.png)
![ID1_seg-axonmyelin](https://user-images.githubusercontent.com/50454820/128307571-1e69fd69-cd0b-438b-81e5-46965bda8e47.png)
I re did the corrections for this image getting rid of all the other grayscale values but I get the same error. Any ideas as to why this is the case, did I miss some colour values?

**Comment by hermancollin** [comment, marked as answer]
Using the last images you provided, I was able to reproduce the error you encountered. Thanks to @mariehbourget, we did some tests and I think we found why the morphometrics computation failed here.

I used GIMP to convert your masks to grayscale mode (they were still in RGB). It turns out this wasn't enough: by default, GIMP does not remove the alpha channel. I suspect Photoshop has the same behavior.

@Paasu please try removing the alpha channel on the grayscale images and see if it works. On our side, the `axondeepseg_morphometrics` script produced the desired output from your images.

**Reply by Paasu** [reply]
I couldn't figure out how to remove alpha channels in Photoshop so I tried it in GIMP and I could remove it. I use the `axondeepseg_morphometrics` line and get the following error however.
[ERROR.txt](https://github.com/neuropoly/axondeepseg/files/6940716/ERROR.txt)

**Reply by hermancollin** [reply]
@Paasu did you convert the image to grayscale? In GIMP you have to set Image > Mode > Grayscale

Also, what is the pixel size you specified?

**Reply by Paasu** [reply]
Yes all the images are grayscale and the alpha channels have been removed. The pixel size I specified 0.001666666666667.

**Reply by hermancollin** [reply]
Unfortunately I am unable to reproduce this. Could you provide the images that gave you this error?

**Reply by Paasu** [reply]
I should have clarified - this occurs when I try to get the ellipse values which is what I am keen on. The circle values come out fine
![ID1_seg-axon](https://user-images.githubusercontent.com/50454820/128405019-97c9ca8e-3dda-4654-94b8-08b4ba4e5e62.png)
![ID1_seg-axonmyelin](https://user-images.githubusercontent.com/50454820/128405024-c8b1e8c9-a164-4d09-861e-382fa45149fb.png)
![ID1_seg-myelin](https://user-images.githubusercontent.com/50454820/128405030-2b5feab0-0cf1-47ed-b60c-03ea57a3604a.png)

**Reply by hermancollin** [reply]
I was able to reproduce the error by specifying the ellipse axon shape. The computation failed because one axon's minor diameter was negative. The warning included in the error you reported gives us the location of the problematic element, approximately (3055,1712). I manually removed the troublesome element (it looked like a segmentation error anyway) and the computation succeeded.
![Screenshot_20210805_163607](https://user-images.githubusercontent.com/83031821/128417844-2173a294-7591-4802-b439-e26be611d0f7.png)

<details><summary>masks</summary>

![paasu_img2_seg-axon](https://user-images.githubusercontent.com/83031821/128417331-a0798bb6-074a-41a8-8a0f-df3a3885f078.png)
![paasu_img2_seg-axonmyelin](https://user-images.githubusercontent.com/83031821/128417333-5cf5d021-fbad-443f-8501-781e5e7700c9.png)
![paasu_img2_seg-myelin](https://user-images.githubusercontent.com/83031821/128417334-5f659a9e-e9f7-4ab4-85f2-2b1bef65105f.png)
</details>

**Reply by Paasu** [reply]
Thank you, getting rid of that element does fix it for this image but is there a way to bypass these errors? The error produced doesn't always point out the problematic element and I can't manually locate the element successfully 100% of the time.

**Reply by jcohenadad** [reply]
@Paasu there are more fundamental issues than this error that you need to address in the first place. Seeing these masks, at least a third of the segmented axons are _wrongly_ segmented (eg: split myelin, yielding wrong "axon" elements). Even if you can get rid of the error, I wouldn't trust the output statistics.

**Reply by hermancollin** [reply]
@Paasu Currently, there is no way to do that. I agree that a more practical behavior for this script would be to output a warning that some computation failed but still finish running morphometrics for all other axons.

As a temporary fix, I suggest to catch the exception. In `AxonDeepSeg/morphometrics/compute_morphometrics.py`, you can change line 163:
```python
stats['gratio'] = (axon_diam / 2) / (axon_diam / 2 + myelin_thickness)
```
by this block of code:
```python
try:
   stats['gratio'] = (axon_diam / 2) / (axon_diam / 2 + myelin_thickness)
except ZeroDivisionError:
   print(f"ZeroDivisionError: skipping element {idx}.")
   stats['gratio'] = 0.0
```

**Reply by Paasu** [reply]
@jcohenadad Initially I was worried by the problem but I was thinking assuming those axon elements as the area of the decompaction might come in handy, especially with my injured dataset.

@hermancollin Thank you! It appears to work, the errors show up as `0` on the outputted excel sheet.
